src/System.Runtime.Extensions/src/System/IO/StreamWriter.cs

```diff
@@ -253,18 +251,29 @@ private void Flush(bool flushStream, bool flushEncoder)
                 }
             }
 
-            int count = _encoder.GetBytes(_charBuffer, 0, _charPos, _byteBuffer, 0, flushEncoder);
-            _charPos = 0;
-            if (count > 0)
+            int count = 0;
+            byte[] byteBuffer = ArrayPool<byte>.Shared.Rent(_encoding.GetMaxByteCount(_charPos));
+            try
+            {
+                count = _encoder.GetBytes(CharBuffer, 0, _charPos, byteBuffer, 0, flushEncoder);
+                _charPos = 0;
+                if (count > 0)
+                {
+                    _stream.Write(byteBuffer, 0, count);
+                }
+            }
+            finally
             {
-                _stream.Write(_byteBuffer, 0, count);
+                ArrayPool<byte>.Shared.Return(byteBuffer);
```

Comment:
Two questions:
1. Have you run any perf tests on this?  I like the change overall, I just want to make sure that the extra rent/return overhead on each flush isn't going to be prohibitive.
2. Do you have a sense for scenarios where this is used how likely it'll be for _charPos to be relatively small?  Seems like if you're auto-flushing, writing line-by-line, etc., it could be relatively small, in which case I'm wondering whether we should conditionally either use span with stack space or rent an array (and while it'd be tempting to use the span Write overload for either, it'd probably be best in this case to conditionally use the appropriate Write).

Reply:
Should we be interlock exchanging the fields storing pooled buffers before we return the buffer to the pool? I know many of our APIs are documented as not thread safe, but the repercussions of returning same buffer multiple times is super high, i.e.
```c#
var buffer = Interlocked.Exchange(ref _byteBuffer, null);
ArrayPool<byte>.Shared.Return(buffer);
```

Reply:
That would only help with concurrent disposal. If we're concerned about incorrect usage causing problems, there are more impactful issues I'd be concerned with, namely disposing while an operation is in progress. To address that we'd need to use Interlockeds to check out and check in the buffer on each usage. That's not cheap.

Reply:
Why do you think that disposing while operation is in progress is more severe? Double return permanently messes up process wide shared pool. Disposing while operation is in progress is a bad problem, but is not permanent (which maybe you are arguing is worse).

Reply:
> we should conditionally either use span with stack space or rent an array

Do you have any feel what is a good size for cut over? i.e. where stackalloc zeroing starts to dominate

Reply:
> Why do you think that disposing while operation is in progress is more severe?

It's not that I think the effect is more severe; it's that I think it's much more likely to happen.  Patterns that would result in concurrently calling Dispose with itself are rare, whereas it'd be pretty easy to accidentally have code like:
```C#
using (var writer = new StreamWriter(...))
{
    writer.WriteAsync(buffer, 0, buffer.Length);
}
```

Reply:
Can we record when we start async operation and then throw in Dispose if there is one outgoing? It would be good diagnostic, at least for the usage bug you provided (as both WriteAsync and Dispose run on the same thread).

Reply:
> Can we record when we start async operation and then throw in Dispose if there is one outgoing?

That's essentially what I'd mentioned about checking out the buffer, then it's just a question of whether that's done with an interlocked or not.

Reply:
> Do you have any feel what is a good size for cut over? i.e. where stackalloc zeroing starts to dominate

I'm hopeful we will have a solution such that there won't be stackalloc zeroing at all.  But even if there is, I'd expect the cost of that zeroing for the small sizes we'd be willing to do this for (maybe 64 chars or something like that) will still be less than the cost of renting/returning from the ArrayPool, though measuring would be the best way to find out.  Anyway, it's just an idea... if it doesn't pan out, that's fine.

Reply:
I think we should be very defensive when we use buffer pooling, at least initially. I would hate to realize in a couple of years that this whole buffer pooling push (which I strongly support) did more damage than good. 

Possibly being paranoid about double dispose takes this philosophy too far, but unless we worry about the perf of interlocking in dispose, maybe it's better to be safe than sorry. Thoughts?

BTW, for check in/out, I would not do interlocked. I am pretty sure that it would show in profiles.

Reply:
> Possibly being paranoid about double dispose takes this philosophy too far, but unless we worry about the perf of interlocking in dispose, maybe it's better to be safe than sorry. Thoughts?

I expect it won't be measurable.  So if it helps us to sleep better, I'm fine with it.  I like sleep :)

Reply:
Added `Interlocked.Exchange` for return; so can be no double return